Modules/Quotas/src/Quotas.Domain/Aggregates/Identities/Identity.cs

```diff
@@ -43,52 +45,48 @@ public async Task UpdateMetrics(IEnumerable<MetricKey> metrics, MetricCalculator
         }
     }
 
-    private void UpdateMetricStatus(MetricKey metricKey, IEnumerable<Quota> quotasFromMetric)
-    {
-        var metricStatus = _metricStatuses.FirstOrDefault(m => m.MetricKey == metricKey);
-        var maxExhaustionDate = quotasFromMetric.Max(q => q.IsExhaustedUntil);
-        if (metricStatus != null)
-        {
-            metricStatus.Update(maxExhaustionDate);
-        }
-        else
-        {
-            _metricStatuses.Add(new(metricKey, maxExhaustionDate));
-        }
-    }
-
     private async Task UpdateMetric(MetricKey metric, IMetricCalculator metricCalculator, CancellationToken cancellationToken)
     {
-        var quotasFromMetric = GetAppliedQuotasForMetric(metric);
-        foreach (var quota in quotasFromMetric)
+        var quotasForMetric = GetAppliedQuotasForMetric(metric);
+        var quotasExhaustion = new Dictionary<Quota, DateTime?>();
+        foreach (var quota in quotasForMetric)
         {
             var newUsage = await metricCalculator.CalculateUsage(
                 quota.Period.CalculateBegin(),
                 quota.Period.CalculateEnd(),
                 Address,
                 cancellationToken);
 
-            quota.UpdateExhaustion(newUsage);
+            quotasExhaustion.Add(quota, quota.CalculateExhaustion(newUsage));
         }
 
-        UpdateMetricStatus(metric, quotasFromMetric);
+        UpdateMetricStatus(metric, quotasExhaustion);
     }
 
     private IEnumerable<Quota> GetAppliedQuotasForMetric(MetricKey metric)
     {
         var allQuotasOfMetric = AllQuotas.Where(q => q.MetricKey == metric);
-        var highestWeight = -1;
+        if (allQuotasOfMetric.Any())
+        {
+            var highestWeight = allQuotasOfMetric.Max(q => q.Weight);
+            var appliedQuotas = allQuotasOfMetric.Where(q => q.Weight == highestWeight).ToArray();
+            return appliedQuotas;
+        }
```

Comment:
You should invert the if to reduce nesting:

```suggestion
        if (!allQuotasOfMetric.Any())
        {
           return Enumerable.Empty<Quota>();
        }
        
        var highestWeight = allQuotasOfMetric.Max(q => q.Weight);
        var appliedQuotas = allQuotasOfMetric.Where(q => q.Weight == highestWeight).ToArray();
        return appliedQuotas;
```